src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java

```diff
@@ -232,27 +99,8 @@ public int getTokenType() {
      * @return underlying node
      */
     @Override
-    public DetailAST getUnderlyingNode() {
+    public Object getUnderlyingNode() {
```

Comment:
lets keep more exact type
https://stackoverflow.com/a/14694885

```
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java b/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java
index 880210d..10415e6 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java
@@ -115,7 +115,7 @@ public class ElementNode extends AbstractElementNode {
      * @return underlying node
      */
     @Override
-    public Object getUnderlyingNode() {
+    public DetailAST getUnderlyingNode() {
         return detailAst;
     }
 
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java b/src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java
index 09b3bad..7d61109 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/xpath/RootNode.java
@@ -99,7 +99,7 @@ public class RootNode extends AbstractRootNode {
      * @return underlying node
      */
     @Override
-    public Object getUnderlyingNode() {
+    public DetailAST getUnderlyingNode() {
         return detailAst;
     }

```

works fine on my local

Reply:
We cannot do this. This goes against the whole purpose of this PR which is to separate XPath from a specific AST. If you look at the original code before my PR, it was already using `DetailAST`. If you wanted to expand xpath to Javadoc node, having Java node classes listed will cause issues.